### DISCUSSION TITLE
Images in a brief presentation not showing up

### DISCUSSION THREAD

**Discussion by pldyang** [jspsych/jsPsych#1111] [Q&A] [answered]
I have an images not showing up problem. The procedure of my experiment is presenting a image and followed by a mask. The shortest duration of images is around 17 ms (1 frame at 60Hz). The overall experiment ran smoothly but sometimes the images (around 15% of the trials) would not be presented at all. It would jump directly from the fixation to the mask. I am not sure what the problem is.
I host the experiment on Pavolvia. Here is [the link to my experiment](https://run.pavlovia.org/AP_lab/pl_scene_face_invert_replicate) and [the link to the code](https://gitlab.pavlovia.org/PL/shared)

To decrease lagging, I preloaded the images (more than 1000) using the following code. The `all_images` variable contains all the images. Here, I only listed the main task part as an example. Nevertheless, the images sometimes are not properly preloaded according to Network tag in the Development tool. 

````html
        var main_target_folder = 'image/main_intact_320/';
        var main_scrambled_folder = 'image/main_mor_320/';
        var main_mask_folder = 'image/main_scr_320/';

        var main_intact_famous_normal = [];
        var main_intact_novel_normal = [];
        var main_scrambled_famous_normal = [];
        var main_scrambled_novel_normal = [];
        var main_mask_intact_fam_normal = [];
        var main_mask_intact_novel_normal = [];

        var main_intact_famous_invert = [];
        var main_intact_novel_invert = [];
        var main_scrambled_famous_invert = [];
        var main_scrambled_novel_invert = [];
        var main_mask_intact_fam_invert = [];
        var main_mask_intact_novel_invert = [];

        for (i = 1; i < 91; i++){
            main_intact_famous_normal[i-1] = main_target_folder + "image" + i + "_fam_normal.png"
            main_intact_novel_normal[i-1] = main_target_folder + "image" + i + "_fake_normal.png"
            main_scrambled_famous_normal[i-1] = main_scrambled_folder + "image" + i + "_fam_normal.png"
            main_scrambled_novel_normal[i-1] = main_scrambled_folder + "image" + i + "_fake_normal.png"
            main_mask_intact_fam_normal[i-1] = main_mask_folder + "image" + i + "_fam_normal.png"
            main_mask_intact_novel_normal[i-1] =  main_mask_folder + "image" + i + "_fake_normal.png"

            main_intact_famous_invert[i-1] = main_target_folder + "image" + i + "_fam_invert.png"
            main_intact_novel_invert[i-1] = main_target_folder + "image" + i + "_fake_invert.png"
            main_scrambled_famous_invert[i-1] = main_scrambled_folder + "image" + i + "_fam_invert.png"
            main_scrambled_novel_invert[i-1] = main_scrambled_folder + "image" + i + "_fake_invert.png"
            main_mask_intact_fam_invert[i-1] = main_mask_folder + "image" + i + "_fam_invert.png"
            main_mask_intact_novel_invert[i-1] =  main_mask_folder + "image" + i + "_fake_invert.png"
        };
        var main_intact_normal = main_intact_famous_normal.concat(main_intact_novel_normal);
        var main_scrambled_normal = main_scrambled_famous_normal.concat(main_scrambled_novel_normal);
        var main_mask_normal = main_mask_intact_fam_normal.concat(main_mask_intact_novel_normal);

        var main_intact_invert = main_intact_famous_invert.concat(main_intact_novel_invert);
        var main_scrambled_invert = main_scrambled_famous_invert.concat(main_scrambled_novel_invert);
        var main_mask_invert = main_mask_intact_fam_invert.concat(main_mask_intact_novel_invert);

        var main_stimuli = [];
        for (i = 0; i < main_intact_normal.length; i++){
            main_stimuli.push({stimulus: main_intact_normal[i], mask: main_mask_normal[i], data: {test_part: 'main', correct_response: 'd', sti_type: 'intact', direction: 'normal'}})
            main_stimuli.push({stimulus: main_scrambled_normal[i], mask: main_mask_normal[i], data: {test_part: 'main', correct_response: 'k', sti_type: 'scrambled', direction: 'normal'}})

            main_stimuli.push({stimulus: main_intact_invert[i], mask: main_mask_invert[i], data: {test_part: 'main', correct_response: 'd', sti_type: 'intact', direction: 'invert'}})
            main_stimuli.push({stimulus: main_scrambled_invert[i], mask: main_mask_invert[i], data: {test_part: 'main', correct_response: 'k', sti_type: 'scrambled', direction: 'invert'}})
        };

        all_images = prac_intact_normal.concat(prac_intact_invert, prac_scrambled_normal, prac_scrambled_invert, prac_mask_normal, prac_mask_invert,
         quest_intact_normal, quest_intact_invert, quest_scrambled_normal, quest_scrambled_invert, quest_mask_normal, quest_mask_invert,
         main_intact_normal, main_intact_invert, main_scrambled_normal, main_scrambled_invert, main_mask_normal, main_mask_invert);

jsPsych.init({
            timeline: timeline,
            preload_images: all_images
        });
````
The code for presenting targets and masks:

````
var main_target = {
        	type: "image-keyboard-response",
        	stimulus: jsPsych.timelineVariable('stimulus'),
        	choices: ['d', 'k'],
        	trial_duration: function(){
                var main_dur = q.mean(); // This is the mean duration obtained from a staircase
                if (main_dur < 17){ 
                    return 17;
                } else {
                    return q.mean();
                }
            },
            data: jsPsych.timelineVariable('data'),
        	on_finish: function(data) {
                var acc = 0;
                if (data.correct_response == jsPsych.pluginAPI.convertKeyCodeToKeyCharacter(data.key_press)) {
                    acc = 1;
                }
                data.accuracy = acc;
                if (q.mean() < 17){ 
                    data.duration = 17;
                } else {
                    data.duration = q.mean();
                }
            }
        };

        var main_mask = {
            type: "image-keyboard-response",
            stimulus: jsPsych.timelineVariable('mask'),
            choices: ['d', 'k'],
            stimulus_duration: null,
            //trial_duration: 1800,
            //response_ends_trial: false,
            data: jsPsych.timelineVariable('data'),
            on_finish: function(data) {
                var acc = 0;
                if (data.correct_response == jsPsych.pluginAPI.convertKeyCodeToKeyCharacter(data.key_press)) {
                    acc = 1;
                }
                data.accuracy = acc;
                data.mask = 1;
            }
        }; 
````

To check the frame dropping issue, I calculated the time_elapsed difference for each trial. Even if the time_elapsed is greater than 17 ms, sometimes the image still doesn't show up. 

Hence, I am not sure whether it's an internet or Pavlovia server problem or I missed anything.

Could anyone help me with this issue? Any comments would be very appreciated.

**Comment by becky-gilbert** [comment]
Hi @pldyang, I wonder if you're hitting browser cache limits, since you're loading so many images. If that's happening then the browser will drop some of your images from the cache, in which case it's likely that some of the images won't be retrieved from the server in time to be shown for such a brief duration.

One thing you can do is check to see if that's happening by opening the network tab in your browser's developer tools and see if the browser is loading any of the files _during_ the experiment (even if they're all successfully preloaded at the start - see [this comment about cache limits](https://github.com/jspsych/jsPsych/discussions/1066#discussioncomment-87255) for more explanation). Another thing you could try is throttling your connection speed in the browser's network tab. If the image dropping is due to a failure to load the images from the cache, then with a slower connection speed you might see that image presentations are dropped more frequently.

If cache limits turn out to be the problem, then you could try preloading the images closer to when they're needed. See this discussion comment/thread about [preloading files at specific points in your task](https://github.com/jspsych/jsPsych/discussions/724#discussioncomment-19258). This discussion thread about [preloading files in the background](https://github.com/jspsych/jsPsych/discussions/961) might also be helpful.

**Reply by pldyang** [reply]
This is really helpful! Sorry that I didn't aware the cache limitation. I did monitor the network tab during the experiment. Sometimes the browser would load `xhr` type of files during the trial but most of them were masks not the target images. 

I will try to load the images closer to the trials and report back whether the problem is fixed or not. Thank you!

**Reply by pldyang** [reply]
I changed the code to preload images before each block and tried different connection speed in the network tab. The overall dropping rate did reduce a little but still average 6 to 8 images were dropped per block (42 images in total). It may just be the presentation time is too short. 
This is a shorter testable version of the experiment: https://run.pavlovia.org/PL/shared/ and the code: https://gitlab.pavlovia.org/PL/shared.

**Reply by becky-gilbert** [reply]
I just tried your task and it looks to me like the preloading is working well - the images are loading at the start and not re-loading during the task. So if some images are still being skipped then the problem is likely due to Josh's point about difficulties getting 100% accuracy for single-frame displays. And as Josh says, you might find that the percentage of dropped images depends on the OS and browser.

**Reply by pldyang** [reply]
Thank you for all the trouble-shooting. Sounds like I should design a more browser friendly design next time!

**Comment by jodeleeuw** [comment, marked as answer]
In addition to @becky-gilbert's important point about the cache, I also want to add that single frame duration accuracy will depend on what browser and OS you are running on. There's no universal method for 100% accuracy of single frame displays via browsers - though we are still looking for one! :grin:

**Reply by pldyang** [reply]
Thank you for the heads up! Yeah I might be pushing the limits of brower performance. Looking forward to the good news! 👍

**Comment by kurokida** [comment]
If you're interested, you can use the plugin I created for psychophysics.
https://jspsychophysics.hes.kyushu-u.ac.jp/

This plugin allows you to specify presentation times in terms of frames.

**Reply by pldyang** [reply]
Thank you for providing this information! I have tried this plugin and it worked great! Nevertheless, we are planning to use Quest staircase algorism in our experiment. So, I wasn't too sure how to implement the Quest algorism with the plugin. I will keep looking into this. :^)
